### ISSUE TITLE
[FR] To extract both matched and non-matched lines or strings to 2 separate files or streams at the same time

### ISSUE THREAD

**Issue by garry-ut99** [Genivia/ugrep#416]
Currently it's needed A or B:
- A) to run ug twice on each file:
`ug -options phrase input_file > output_file_A_.txt` (extracts matched lines)
`ug -options -v phrase input_file > output_file_B.txt` (extract non-matched/inverted lines)
- B) or to run two separate instances of ug at the same time

**A:** is 2x slower, as it needs to run ag two times

**B:** 
- takes 2x more RAM as it loads the same file twice into RAM
- also incerases CPU usage 2x, with an additional risk of CPU being bottlenecked and causing speed reduction, 
expecially when processing heavy files or big ammout of files and doing other tasks on PC meanwhile

I know there is possible to redirect both `stdout` and `stderr` to separate files in cmd line: `command 2>> error 1>> output` as per https://stackoverflow.com/questions/7901517/how-to-redirect-stderr-and-stdout-to-different-files-in-the-same-line-in-script thought I'm not sure whether there is possible to redirect `stdout` to 2 separate files interchangeably from a cmd line, if not, then output / file redirection would have to be handled by ugrep itself rather than cmd or maybe `stderr` could be used as an additional output stream while using the feature.

Also if the output redirections would be handled by ugrep internally, then it would be good to have an option to add patterns to output filenames, so each processed file has two corresponding output files, whose names are input_filename+patternA and input_filename+patternB, for example: ug processes input file `file.txt`, then if provided patterns are `_A` and `_B` , the output filenames are: `file_A.txt` and `file_B.txt`, useful when processing large ammout of files in different folders.

(Keep in mind, that despite it's called "feature request", this is just a proposition / idea from my side, I don't force you to implement this idea, you decide for yourself).

**Comment by genivia-inc** [comment]
This can simply be done with one `ug` call using option `-y` (any line/passthrough), because the separator is `-` for context lines:
```console
$ ugrep -yhn -options phrase input_file > temp_file
$ ugrep -P '^[0-9]+:(.*)' --format='%1%~' temp_file > output_file_A
$ ugrep -P '^[0-9]+-(.*)' --format='%1%~' temp_file > output_file_B
```
The second `ugrep -P` strips the line numbers and separator (a `:` or a `-`).

(Note: this won't work with  some options you pick like `-o` that conflicts with `-y`.)

Doing any of this only makes any sense if the `phrase` pattern is very large or complex, like consuming a file with tens of thousands of patterns. Normally ugrep uses only limited memory with a 256KB sliding window over the input so the size of the `input_file` does not matter.

For example, searching 1000 words doesn't use much RAM or time. These are the 1000 words from the [ugrep benchmark](https://github.com/Genivia/ugrep-benchmarks)) and searching a 100MB file:
```
/usr/bin/time -l ugrep -c -F -f benchmarks/words/4.txt benchmarks/corpi/enwik8 --stats=vm
13975

Searched 1 file in 0.186 seconds: 1 matching (100%)
Searched 1128023 lines: 13975 matching (1.239%)
The following pathname selections and search constraints were applied:
  --fixed-strings
  --no-hidden (default)
Lines matched if:
  a string in benchmarks/words/4.txt matches
VM: 8256 nodes (0ms) 8256 edges (0ms) 17513 opcode words (0ms)
        0.19 real         0.18 user         0.01 sys
             7110656  maximum resident set size
                   0  average shared memory size
                   0  average unshared data size
                   0  average unshared stack size
                 532  page reclaims
                   0  page faults
                   0  swaps
                   0  block input operations
                   0  block output operations
                   0  messages sent
                   0  messages received
                   0  signals received
                   0  voluntary context switches
                   3  involuntary context switches
          1132759035  instructions retired
           561118827  cycles elapsed
             5243584  peak memory footprint
```
The VM stats show the internal virtual machine size in opcode words as well as the DFA size in nodes and edges.

**Comment by genivia-inc** [comment]
Wwith the `tee` utility you can do this in parallel:
```
ugrep -yhn -options phrase input_file | tee temp_file | ugrep -P '^[0-9]+:(.*)' --format='%1%~' > output_file_A
ugrep -P '^[0-9]+-(.*)' --format='%1%~' temp_file > output_file_B
```
For fully task-parallelism, use a named pipe instead of `temp_file`. The `tee` will send to it and the last ugrep call will read it.

**Comment by garry-ut99** [comment]
> genivia-inc : https://github.com/Genivia/ugrep/issues/416#issuecomment-2276584124 (...) Doing any of this only makes any sense if the phrase pattern is very large or complex (...)

Which means it's useless in a typical scenario which I asked: when using a single text word as a phrase, as it's x1.72 - x2.73 slower than a typical sequential method (A) (benchmarked on a 661 MB text file with 33 000 000 domains - the same file we were discussing in the other thread).

> For example, searching 1000 words doesn't use much RAM or time.

661MB file, 33 000 000 lines (domains)

As for RAM, a single ug instance took up to 400MB for a while (peak), two of them at once can take 800MB, (I didn't test on larger files like 6GB or 16GB)

As for time (includes removing temp file):
- a single word: ~500 hits, method A (sequential) : 3.23s ;  your method 8.82s
- a single word: 16 600 000 hits, method A (sequential) : 6.67s ;  your method 11.45s

Also as for disk space:
- your method creates additional temp file which can be 1.5x larger than input file, for example almost 1GB for a 660MB test file, because it adds `:` + `-` + numbers on each line which grows quickly and take much additional space, it wastes disk's lifespan, especially having dozens of GB of files to process, as it generates additional 150% data on disk, atop of the 100% of sourcel data + 100% of output data, also finally the temp file has to be also removed.

> genivia-inc :  https://github.com/Genivia/ugrep/issues/416#issuecomment-2276596216 : Wwith the tee utility you can do this in parallel:

I'll test it a bit later, as I'm busy.

**Comment by genivia-inc** [comment]
> Also as for disk space:
> 
> * your method creates additional temp file which can be 1.5x larger than input file (or even 2x-3x, depending on input file content), for example almost 1GB for a 660MB test file, because it inserts `:` + `-` and also numbers in each line, which grows quickly and take much additional space, it wastes disk's lifespan, especially having dozens of GB of files to process, as it generates additional 150% data on disk (or even 200%-300% depending on input file's content), atop of the 100% of source data + 100% of output data, also finally the temp file has to be also removed.

Ahem... that's why a named pipe is best. No file writes at all. Please read.

I think this answers your question.

**Comment by garry-ut99** [comment]
> You didn't appear to get my point. 

Maybe because you weren't clear enought, since you created two separate comments, it misled me into thinking that you wanted me to try the both methods separately (non-parallel and parallel), perhaps instead of blaming me immediately, you should have been edit your previous comment just like you did before, instead of creating another one, or to make it very clear that the both comments are directly connected. Optionally, I can agree that the blame for the context/wording misunderstanding lies on both sides.

> Please read.

Please write clearly and don't blame. Optionally, I can agree that the blame lies on both sides, I also do understand you didn't like my reply, as it was created with a misassumption, and puts your method in a bad light, anyway your "please read" followed by closing thread, reminds me "please open a new issue" followed by locking thread, from the other rude guy, the difference between him and you is that he was rude since the beginning:
- https://github.com/BurntSushi/ripgrep/discussions/2867



> I think this answers your question.

Maybe, but theory doesn't always match reality, on my side I can be sure only after doing benchmarks, when I get time, but I don't know whether I'm interested anymore.

**Comment by genivia-inc** [comment]
I've deleted the last message posted by @garry-ut99 because it was not in compliance with the [code of conduct](https://github.com/Genivia/ugrep/blob/master/CODE_OF_CONDUCT.md).

**Comment by genivia-inc** [comment]
To prove it to you, my suggestion was to use `tee` and a named pipe and put in a script file e.g. `teeters.sh` (my example here uses 1000 words to search and a 100MB file to search):
```console
mkfifo teepipe
ugrep -y -n -F -f words4.txt enwik8 | tee teepipe | ugrep -P '^[0-9]+:(.*)' --format='%1%~' > output1.txt & \
ugrep -P '^[0-9]+-(.*)' --format='%1%~' < teepipe > output2.txt
```
then this takes 0.33 seconds.
```console
time ./teetest.sh
0.490u 0.160s 0:00.33 196.9%	0+0k 0+0io 0pf+0w
              ^^^^^^^
```
when a single ugrep call takes 036 seconds:
```console
time ugrep -y -n -F -f words4.txt enwik8 > output1.txt
0.290u 0.068s 0:00.36 97.2%	0+0k 0+0io 8pf+0w
              ^^^^^^^
```
The `tee` and two extra `ugrep -P` aren't expensive to execute and use less memory than the main work done by the first `ugrep` when the search pattern is very large (an important assumption!)

This runs 2x faster than executing `ugrep` and `ugrep -v` separately in about 0.36s + 0.36s = 0.72 seconds to get the results for `output1.txt` and `output2.txt` separately. You could run `ugrep` and `ugrep -v` in parallel in 0.36 seconds, but if the search pattern is very large then this consumes 2x the memory.

**Comment by garry-ut99** [comment]
### 1

> garry-ut99: (unjustly removed comment) : https://github.com/Genivia/ugrep/issues/416#issuecomment-2278364348 ==> https://github.com/Genivia/ugrep/issues/416#event-13824422863 :
> 
> Maybe because you weren't clear enought, since you created two separate comments, it misled me into thinking that you wanted me to try the both methods separately (non-parallel and parallel), perhaps instead of blaming me immediately, you should have been edit your previous comment just like you did before, instead of creating another one, or to make it very clear that the both comments are directly connected. Optionally, I can agree that the blame for the context/wording misunderstanding lies on both sides.
> 
> > Please read.
> 
> Please write clearly and don't blame. Optionally, I can agree that the blame lies on both sides, I also do understand you didn't like my reply, as it was created with a misassumption, and puts your method in a bad light, anyway your "please read" followed by closing thread, reminds me "please open a new issue" followed by locking thread, from the other rude guy, the difference between him and you is that he was rude since the beginning:
> 
> - https://github.com/BurntSushi/ripgrep/discussions/2867
> 
> > I think this answers your question.
> 
> Maybe, but theory doesn't always match reality, on my side I can be sure only after doing benchmarks, when I get time, but I don't know whether I'm interested anymore.
> 

> genivia-inc : https://github.com/Genivia/ugrep/issues/416#issuecomment-2278534747 : I've deleted the last message posted by @garry-ut99 because it was not in compliance with the [code of conduct](https://github.com/Genivia/ugrep/blob/master/CODE_OF_CONDUCT.md?rgh-link-date=2024-08-09T18%3A42%3A03Z).

So I've repasted it here above, because I don't see anything offensive in my previous comment, which could break the "code of conduct", only few words of truth, in the worst case, I only adapted my level to your level, so what is your problem then, it's not me who started.

You only provided a generic link to "code of conduct" without actually providing and proving which specific sentences of my previous comment break which specific points of "code of conduct", you proved nothing, yet you still removed my comment, accusing me of breaking "code of conduct". The irony is that it's your and the other guy's actions, which can be considered breaking of "code of conduct", how can you accuse one of breaking "code of coduct" after previously breaking it by yourself.

> genivia-inc :  Edit: screenshot evidence:

Since you're speaking of evidence, where is your evidence that I broke "code of conduct" in my previous comment..., you seem to selectively choose when to provide evidence, and when to not..., before removing my comment again, prove anything, of course allow me to reply to it (to agree or disagree with your "evidence"), otherwise it will be nothing more than desperate and deliberate trolling / censorship, and that are desperately out of arguments, and I'll leave this place forever, go ahead, if this is what you want, I don't care.

<hr>

### 2

> genivia-inc : https://github.com/Genivia/ugrep/issues/416#issuecomment-2278749375 : To prove it to you, my suggestion was to use `tee` and a named pipe and put in a script file e.g. `teetest.sh`

To prove waht exactly....you can prove your algorithm, but certainly not the drama part, as there was no such suggestion in your first comment, as it doesn't contain words: "tee" or "pipe"  (CTRL + F in browser returns nothing), you only suggested it in your second comment, which I didn't realise it was directly connected/related to the previous comment, not to say I didn't even had time back then to try method proposed in your second comment, as I was tired and busy already.

<hr>

### 3

> genivia-inc : https://github.com/Genivia/ugrep/issues/416#issuecomment-2278749375 : my suggestion was to use `tee` and a named pipe and put in a script file e.g. `teetest.sh`

I didn't read the thread yesterday (I did a little break of reading), so I didn't see your last comment yesterday, nor that you removed my comment, I see it only now, however I spent yesterday few hours and was testing your method proposed in your second comment, also today I aslo read your last comment, but unfortunately you proposed me solution for Linux, and I failed when trying to port your Linux command to Windows batch version (I also heard about PowerShell way), so if you want me to confirm that your method works on Windows, then maybe you can fix it (or not, I don't care, it's your choice), I spent enought time on this:
- Tee for Windows: https://github.com/dEajL3kA/tee-win32
- windows-named-pipe-utils : https://github.com/psmay/windows-named-pipe-utils

<details>
<summary>input_to_output.bat (SPOILER):</summary>

```
@echo off  
set startTime=%time%
REM main path to RAMDISK
set path=F:
set path_ug=ugrep-windows-x64
set path_tee=_input_to_output\tee-win32.2023-11-27
set path_pipes=_input_to_output\windows-named-pipe-utils
REM input_file size 661 MB / 32237207 lines / each line is domain name / sorted a-z order / duplicates removed
set input_file=input_file.txt
set temp_file=temp_file.txt
set output_file_A=output_file_A.txt
set output_file_B=output_file_B.txt
set phrase="someone"
set matched="^[0-9]+:(.*)"
set unmatched="^[0-9]+-(.*)"



REM METHOD 1
REM Standard synchronic method for Windows: 3.28s total 
REM using --encoding=UTF-8 and -a, otherwise it throws error: "binary file"
REM %path_ug%\ug --encoding=UTF-8 -ahrP --stats  %phrase% "%path%\%input_file%" > "%path%\%output_file_A%"
REM Searched 1 file in 0.59 seconds with 7 threads: 1 matching (100%)
REM Searched 32237206 lines: 509 matching (0.001579%)
REM %path_ug%\ug --encoding=UTF-8 -ahrP -v --stats  %phrase% "%path%\%input_file%" > "%path%\%output_file_B%"
REM Searched 1 file in 2.69 seconds with 7 threads: 1 matching (100%)
REM Searched 32237207 lines: 32236698 matching (100%)




REM METHOD 2 (just some test)
REM New method for Windows (test run - partial (A)): 4.42s total, it doesn't seem to work fast if Tee writes to a file
REM it needs to write to a named pipe, but I failed at "New method for Windows"
%path_ug%\ugrep --encoding=UTF-8 -ayhn --stats -P %phrase% "%path%\%input_file%" | %path_tee%\tee-x64.exe %path%\%temp_file% | %path_ug%\ugrep --stats -P %matched% --format="%1" > "%path%\%output_file_A%"
REM DEL /Q %path%\%temp_file%




REM METHOD 3
REM New method for Windows (attempt to convert METHOD 4 from Linux examples to Windows, and to use named pipes): 
REM maybe can also use %path_tee%\tee-x64.exe, for example:
REM "ugrep | %path_tee%\tee-x64.exe %path_pipes%\createAndWritePipe MyPipeName" but it throws some i/o errors

REM start pipe server (need to remove REM):
REM %path_ug%\ugrep --encoding=UTF-8 -ayhn --stats -P %phrase% "%path%\%input_file%" | %path_pipes%\createAndWritePipe MyPipeName

REM pipe clients won't run because server is blocking processing the current file, and expects clients to be run as separate consoles
REM (via start "" /b), which complicates things, doesn't work as expected, is too slow anyway:
REM %path_ug%\ugrep --stats -P %matched% --format="%1" < \\.\pipe\MyPipeName > "%path%\%output_file_A%"
REM %path_ug%\ugrep --stats -P %unmatched% --format="%1" < \\.\pipe\MyPipeName > "%path%\%output_file_B%"

echo Start Time: %startTime%
echo Finish Time: %time%
PAUSE




REM METHOD 4 (examples from Linux)
REM New method for Linux (examples):

REM ugrep -yhn -options phrase input_file | tee temp_file | ugrep -P '^[0-9]+:(.*)' --format='%1%~' > output_file_A
REM ugrep -P '^[0-9]+-(.*)' --format='%1%~' temp_file > output_file_B

REM mkfifo teepipe
REM ugrep -y -n -F -f words4.txt enwik8 | tee teepipe | ugrep -P '^[0-9]+:(.*)' --format='%1%~' > output1.txt & \
REM ugrep -P '^[0-9]+-(.*)' --format='%1%~' < teepipe > output2.txt
```


</details>

**Comment by genivia-inc** [comment]
@garry-ut99  Your replies are in violation of the code of conduct, which states, among other things:

- Using welcoming and inclusive language
- Being respectful of differing viewpoints and experiences
- Gracefully accepting constructive criticism

You replies do not align with all three.

**Comment by garry-ut99** [comment]
> genivia-inc : Your replies are in violation of the code of conduct, which states, among other things:
> genivia-inc : Your replies do not align with all three.

You've quoted ponts from your "code of conduct" but you still didn't quote sentences / parts of my comments, which are supposed to violate these points, hence your accusations are unfounded. Not to mention you violate these points by yourself, and you accuse me of doing the same at the same time, this is hypocrisy and deliberate trolling.  I already have defeated you on arguments, you didn't provide any meaningful reply or aguments, except blaming me, evading truth, manipulating truth and removing my comments over and over, you are a perfect example of a troll.

>  genivia-inc : then please find help elsewhere.

I didn't ask you for help, you need to seek help from a doctor.

>   genivia-inc : All the best.

You too, bye.
